modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,346 @@
+"""
+Take in bounding box coordinates from Geolocation and use to estimate landing pad locations.
+Returns an array of classes, each containing the x coordinate, y coordinate, and spherical 
+covariance of each landing pad estimation.
+"""
+
+import numpy as np
+import sklearn
+import sklearn.datasets
+import sklearn.mixture
+
+from .. import object_in_world
+from .. import detection_in_world
+
+
+class ClusterEstimation:
+    """
+    Estimate landing pad locations based on landing pad ground detection. Estimation
+    works by predicting 'cluster centres' from groups of closely placed landing pad
+    detections.
+
+    ATTRIBUTES
+    ----------
+    min_activation_threshold: int
+        Minimum total data points before model runs.
+
+    min_new_points_to_run: int
+        Minimum number of new data points that must be collected before running model.
+
+    random_state: int
+        Seed for randomizer, to get consistent results.
+
+    METHODS
+    -------
+    run()
+        Take in list of landing pad detections and return list of estimated landing pad locations
+        if number of detections is sufficient, or if manually forced to run.
+
+    __decide_to_run()
+        Decide when to run cluster estimation model.
+
+    __sort_by_weights()
+        Sort input model output list by weights in descending order.
+
+    __convert_detections_to_point()
+        Convert DetectionInWorld input object to a [x,y] position to store.
+
+    __filter_by_points_ownership()
+        Removes any clusters that don't have any points belonging to it.
+
+    __filter_by_covariances()
+        Removes any cluster with covariances much higher than the lowest covariance value.
+    """
+    __create_key = object()
+
+    # VGMM Hyperparameters
+    __COVAR_TYPE = "spherical"
+    __MODEL_INIT_PARAM = "k-means++"
+    __WEIGHT_CONCENTRATION_PRIOR = 100
+    __MEAN_PRECISION_PRIOR = 1E-6
+    __MAX_MODEL_ITERATIONS = 1000
```

Comment:
What happens when the model fails to converge? Does it raise an exception or return a value or something else?

Reply:
the model will raise a warning that won't interrupt the execution of the code. However, failure to converge does mean that the model's prediction won't be accurate since it did not find a set of parameters to fit the data well. From the scikit-learn website: 

The method fits the model n_init times and sets the parameters with which the model has the largest likelihood or lower bound. Within each trial, the method iterates between E-step and M-step for **max_iter** times until the change of likelihood or lower bound is less than **tol**, otherwise, a **ConvergenceWarning** is raised.

tol:
The convergence threshold. EM iterations will stop when the lower bound average gain is below this threshold.

I can also access the convergence state by using vgmm.converged_, getting True or False depending on its state.